### PR TITLE
Add Authentication to frontend and backend

### PR DESCRIPTION
**Pull request by ragizaki**
This PR adds authentication to the frontend and the backend using Flask's session and React's Context features.

BACKEND:
To check if a user is logged in on the backend, you can access the global `session` dict, with `isLoggedIn` or `user` fields. There are two new backend routes, `/api/login` and `/api/logout` which are responsible for changing the session variable accordingly.

FRONTEND:
The frontend now has an `AuthContext` which stores a user's username, or `null` if they are not logged in. To access the current user, you can use
```javascript
import useAuth from 'hooks/useAuth`
const { user } = useAuth()
// user will either be null or a valid username, so a quick check can be done to see if a user is logged in
```

TODO:
Currently when a user signs up, they are redirected to the login page. In the future, we can make this auto login if we want.

## Acceptance Criteria
- After logging in, the user is saved in localStorage, and the navbar shows the Home, Create Post and Logout links
- After logging out, the user is removed from localStorage, and the navbar shows only the Home, Sign Up and Login links
- The backend session is empty when a user is not logged in, and has the `isLoggedIn` and `user` fields if a user is logged in

### CODE REVIEW

File: app/src/context/AuthContext.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      return state;
+  }
+};
+
+export const AuthContextProvider = ({ children }) => {
+  const [state, dispatch] = useReducer(authReducer, {
+    user: localStorage.getItem("user"),
+  });
+
+  console.log("AuthContext state: ", state);
```
Comment (raveen-singh) [inline app/src/context/AuthContext.js]: can you remove this log

**Review by raveen-singh — COMMENTED** [review]
The FE looks good, but i'll leave the BE review to someone else since I worked on it.

**Comment by jennybai01** [comment]
looks good, i think it will be useful to show a "logged in as {user}". also hwo do u get current user from backend?

**Comment by raveen-singh** [comment]
> looks good, i think it will be useful to show a "logged in as {user}". also hwo do u get current user from backend?

Do you want that in a separate function, or to be returned when the user logs in?

**Comment by jennybai01** [comment]
i mean for it to be shown on every page (in navbar?) as long as user is logged in. idk if thats difficult or not

**Comment by raveen-singh** [comment]
> i mean for it to be shown on every page (in navbar?) as long as user is logged in. idk if thats difficult or not

oh gotcha, that should be possible. @ragizaki lmk if you can add it in without it being a lot of work. We can close this PR after .

**Pull request merged**